### PR TITLE
Fix unmarshal's detection of invalid fields in JSON input.

### PR DESCRIPTION
**Pull request by robshakir**
Fixes a bug discovered during schema validation. This change may impact
existing validation checking which was overly lax.

### CODE REVIEW

File: ytypes/util_schema_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}, {
+		desc: "hierarchical fields, populated",
+		inJSONTree: map[string]interface{}{
+			"config": map[string]interface{}{
+				"description": "hello-world",
+			},
+		},
+		inDataPaths: [][]string{
+			{"config", "description"},
+		},
```
Comment (wenovus) [inline ytypes/util_schema_test.go:60]: Could you add tests for when the JSON path is longer or shorter than the datapath?

e.g.
Is "config" as a leaf a valid path for [config description]?
Is "config/description/name" a valid path for [config description]?

File: ytypes/schema_tests/testdata/unexpected-ntp.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+                            "offset": "0",
+                            "peer-type": "UNICAST",
+                            "poll-interval": 1024,
+                            "reach": 0,
+                            "refid": "INIT",
+                            "root-delay": 0,
+                            "root-dispersion": "0",
+                            "stratum": 16,
+                            "tally-code": "REJECT",
+                            "when": -484302876
```
[the remark this answers is not in the document]
Comment (robshakir) [reply ytypes/schema_tests/testdata/unexpected-ntp.json:18]: Agreed that it's not clear here - but I'm not quite sure how to annotate it this way (this was a real-world scenario that is being used as an integration test) -- we can't include comments in JSON files AFAIK, and it seems rather counter to the idea of an integration test to add something in the payload here that changes the test.

Do you have a suggestion on how to make this clearer? We could potentially add something in a comment in the `.go` file.

File: ytypes/util_schema.go [hunk 3]
```diff
@@ -206,24 +207,55 @@ func removeNonDataPathElements(parentSchema, schema *yang.Entry, paths [][]strin
...
-// checkDataTreeAgainstPaths checks each of dataPaths against the first level
-// of the data tree. It returns an error with the first element in the data tree first
-// level that is not found in dataPaths.
-// This function is used to verify that the jsonTree does not contain any elements
-// in the first level that do not have data paths found in the schema.
+// checkDataTreeAgainstPaths checks that all paths that are defined in jsonTree are valid according
+// to the paths that are supplied in dataPaths. In the case that the slice of dataPaths contains elements
+// that consist of more than one path (e.g., [config description]) the paths are recursively checked. This
+// ensures that where there are elements that are compressed out of the data tree, then they are validated.
+// It returns an error if there are fields that are in the JSON that are not specified in the dataPaths.
```
[the remark this answers is not in the document]
Comment (robshakir) [reply ytypes/util_schema.go:214]: Reworded to reflect what you said here with an example, thanks.

File: ytypes/util_schema.go [hunk 4]
```diff
@@ -206,24 +207,55 @@ func removeNonDataPathElements(parentSchema, schema *yang.Entry, paths [][]strin
...
 func checkDataTreeAgainstPaths(jsonTree map[string]interface{}, dataPaths [][]string) error {
-	// Go over all first level JSON tree map keys to make sure they all point
-	// to valid schema paths.
-	pm := map[string]bool{}
-	for _, sp := range dataPaths {
-		pm[util.StripModulePrefix(sp[0])] = true
+	// Primarily, we build a tree that consists of all the valid paths that we were provided
```
[the remark this answers is not in the document]
Comment (robshakir) [reply ytypes/util_schema.go:216]: Fixed. :-)

File: ytypes/util_schema_test.go (hunk 1 above)
Comment (robshakir) [reply ytypes/util_schema_test.go:60]: Sure -- thanks for catching these cases. Added the test cases.

**Comment by robshakir** [comment]
Thanks for the review, PTAL!

File: ytypes/schema_tests/testdata/unexpected-ntp.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenovus) [reply ytypes/schema_tests/testdata/unexpected-ntp.json:18]: Maybe put the extra name in the name of the file? Interesting I didn't know JSON didn't support comments.

**Review by wenovus — APPROVED** [review]

File: ytypes/util_schema.go [hunk 5]
```diff
@@ -206,24 +207,58 @@ func removeNonDataPathElements(parentSchema, schema *yang.Entry, paths [][]strin
...
-// level that is not found in dataPaths.
-// This function is used to verify that the jsonTree does not contain any elements
-// in the first level that do not have data paths found in the schema.
+// checkDataTreeAgainstPaths checks that all paths that are defined in jsonTree match at least one of
+// the paths that are supplied in dataPaths. A match exists when a path prefix of a jsonTree element is
+// equal to one or more of the given dataPaths. Since each dataPath is checked as a valid prefix, the
+// maximum depth of the check is limited to the length of the dataPaths specified. For example, if the jsonTree
+// contains an element which has the path /foo/bar/baz, and dataPaths specifies /foo, then only /foo is checked
+// to be a valid path, no assertions are made about the validity of 'bar' as a child of 'foo'.
+// are valid according
```
[the remark this answers is not in the document]
Comment (robshakir) [reply ytypes/util_schema.go:216]: Fixed, thanks.

File: ytypes/schema_tests/testdata/unexpected-ntp.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (robshakir) [reply ytypes/schema_tests/testdata/unexpected-ntp.json]: I simplified the example and renamed the file -- thanks!

**Pull request merged**